src/Lykke.Service.LykkeAdapter.Core/Handlers/OrderBookHandlerDecorator.cs

```diff
@@ -0,0 +1,130 @@
+﻿using Lykke.Service.LykkeAdapter.Core.Domain.OrderBooks;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Common.Log;
+using Lykke.Job.OrderBooksCacheProvider.Client;
+
+namespace Lykke.Service.LykkeAdapter.Core.Handlers
+{
+    public class OrderBookHandlerDecorator : IHandler<LykkeOrderBook>
+    {
+        private const string Name = "lykke";
+        private readonly IHandler<TradingOrderBook> _rabbitMqHandler;
+        private readonly IOrderBookProviderClient _bookProviderClient;
+        private readonly ILog _log;
+        private readonly Dictionary<string, LykkeOrderBook> _halfOrderBooks = new Dictionary<string, LykkeOrderBook>();
+
+        public OrderBookHandlerDecorator(IHandler<TradingOrderBook> rabbitMqHandler, IOrderBookProviderClient bookProviderClient, ILog log)
+        {
+            _rabbitMqHandler = rabbitMqHandler;
+            _bookProviderClient = bookProviderClient;
+            _log = log;
+        }
+
+        public async Task Handle(LykkeOrderBook message)
+        {
+            var currentKey = message.AssetPair + message.IsBuy;
+            var wantedKey = message.AssetPair + !message.IsBuy;
+
+            await CheckInited(message, currentKey, wantedKey);
+
+            _halfOrderBooks[currentKey] = message;
+            if (!_halfOrderBooks.TryGetValue(wantedKey, out var otherHalf))
+            {
+                otherHalf = new LykkeOrderBook()
+                {
+                    AssetPair = message.AssetPair,
+                    IsBuy = !message.IsBuy,
```

Comment:
Is it correct?

Reply:
yes here we create empty opposit Half

Reply:
I briefly tested it locally and it looked OK. Alexey said he tested it on dev it worked as expected and said it was ok to merge it. Apparently the solution here is to fetch missing asks/bids from this orderbook cache provider.